// zca-js/dist/cjs/models/Message.cjs
'use strict';

var context = require('../context.cjs');

exports.MessageType = void 0;
(function (MessageType) {
    MessageType[MessageType["DirectMessage"] = 0] = "DirectMessage";
    MessageType[MessageType["GroupMessage"] = 1] = "GroupMessage";
})(exports.MessageType || (exports.MessageType = {}));
class Message {
    constructor(data) {
        this.type = exports.MessageType.DirectMessage;
        this.data = data;
        this.threadId = data.uidFrom == "0" ? data.idTo : data.uidFrom;
        this.isSelf = data.uidFrom == "0";
        if (data.idTo == "0")
            data.idTo = context.appContext.uid;
        if (data.uidFrom == "0")
            data.uidFrom = context.appContext.uid;
    }
}
class GroupMessage {
    constructor(data) {
        this.type = exports.MessageType.GroupMessage;
        this.data = data;
        this.threadId = data.idTo;
        this.isSelf = data.uidFrom == "0";
        if (data.uidFrom == "0")
            data.uidFrom = context.appContext.uid;
    }
}

exports.GroupMessage = GroupMessage;
exports.Message = Message;
