// zca-js/dist/cjs/apis/deleteMessage.cjs
'use strict';

var context = require('../context.cjs');
var ZaloApiError = require('../Errors/ZaloApiError.cjs');
var Message = require('../models/Message.cjs');
var utils = require('../utils.cjs');

function removeMessageFactory(api) {
    const URLType = {
        [Message.MessageType.DirectMessage]: `${api.zpwServiceMap.chat[0]}/api/message/delete?zpw_ver=${context.appContext.API_VERSION}&zpw_type=${context.appContext.API_TYPE}`,
        [Message.MessageType.GroupMessage]: `${api.zpwServiceMap.group[0]}/api/group/deletemsg?zpw_ver=${context.appContext.API_VERSION}&zpw_type=${context.appContext.API_TYPE}`,
    };
    /**
     * Delete a message
     *
     * @param message Message or GroupMessage instance
     * @param onlyMe Delete message for only you
     *
     * @throws ZaloApiError
     */
    return async function deleteMessage(message, onlyMe = true) {
        var _a;
        if (!context.appContext.secretKey)
            throw new ZaloApiError.ZaloApiError("Secret key is not available");
        if (!context.appContext.imei)
            throw new ZaloApiError.ZaloApiError("IMEI is not available");
        if (!context.appContext.cookie)
            throw new ZaloApiError.ZaloApiError("Cookie is not available");
        if (!context.appContext.userAgent)
            throw new ZaloApiError.ZaloApiError("User agent is not available");
        if (!(message instanceof Message.Message) && !(message instanceof Message.GroupMessage))
            throw new ZaloApiError.ZaloApiError("Expected Message or GroupMessage instance, got: " + ((_a = message === null || message === void 0 ? void 0 : message.constructor) === null || _a === void 0 ? void 0 : _a.name));
        if (message.isSelf && onlyMe === false)
            throw new ZaloApiError.ZaloApiError("To delete your message for everyone, use undo api instead");
        const params = {
            toid: message instanceof Message.Message ? message.threadId : undefined,
            grid: message instanceof Message.GroupMessage ? message.threadId : undefined,
            cliMsgId: Date.now(),
            msgs: [
                {
                    cliMsgId: String(message.data.cliMsgId),
                    globalMsgId: String(message.data.msgId),
                    ownerId: String(message.data.uidFrom),
                    destId: String(message.threadId),
                },
            ],
            onlyMe: onlyMe ? 1 : 0,
            imei: message instanceof Message.Message ? context.appContext.imei : undefined,
        };
        utils.removeUndefinedKeys(params);
        const encryptedParams = utils.encodeAES(context.appContext.secretKey, JSON.stringify(params));
        if (!encryptedParams)
            throw new ZaloApiError.ZaloApiError("Failed to encrypt message");
        const response = await utils.request(URLType[message.type], {
            method: "POST",
            body: new URLSearchParams({
                params: encryptedParams,
            }),
        });
        const result = await utils.handleZaloResponse(response);
        if (result.error)
            throw new ZaloApiError.ZaloApiError(result.error.message, result.error.code);
        return result.data;
    };
}

exports.removeMessageFactory = removeMessageFactory;
