// zca-js/dist/cjs/apis/getUserInfo.cjs
'use strict';

var context = require('../context.cjs');
var ZaloApiError = require('../Errors/ZaloApiError.cjs');
var utils = require('../utils.cjs');

function getUserInfoFactory(api) {
    const serviceURL = utils.makeURL(`${api.zpwServiceMap.profile[0]}/api/social/friend/getprofiles/v2`, {
        zpw_ver: context.appContext.API_VERSION,
        zpw_type: context.appContext.API_TYPE,
    });
    /**
     * Get user info using user id
     *
     * @param userId user id
     *
     * @throws ZaloApiError
     */
    return async function getUserInfo(userId) {
        if (!context.appContext.secretKey)
            throw new ZaloApiError.ZaloApiError("Secret key is not available");
        if (!context.appContext.imei)
            throw new ZaloApiError.ZaloApiError("IMEI is not available");
        if (!context.appContext.cookie)
            throw new ZaloApiError.ZaloApiError("Cookie is not available");
        if (!context.appContext.userAgent)
            throw new ZaloApiError.ZaloApiError("User agent is not available");
        if (!context.appContext.language)
            throw new ZaloApiError.ZaloApiError("Language is not available");
        if (!userId)
            throw new ZaloApiError.ZaloApiError("Missing user id");
        const params = {
            phonebook_version: context.appContext.extraVer.phonebook,
            friend_pversion_map: [`${userId}_0`],
            avatar_size: 120,
            language: context.appContext.language,
            show_online_status: 1,
            imei: context.appContext.imei,
        };
        const encryptedParams = utils.encodeAES(context.appContext.secretKey, JSON.stringify(params));
        if (!encryptedParams)
            throw new ZaloApiError.ZaloApiError("Failed to encrypt params");
        const response = await utils.request(serviceURL, {
            method: "POST",
            body: new URLSearchParams({
                params: encryptedParams,
            }),
        });
        const result = await utils.handleZaloResponse(response);
        if (result.error)
            throw new ZaloApiError.ZaloApiError(result.error.message, result.error.code);
        return result.data;
    };
}

exports.getUserInfoFactory = getUserInfoFactory;
