// zca-js/dist/cjs/apis/undo.cjs
'use strict';

var context = require('../context.cjs');
var ZaloApiError = require('../Errors/ZaloApiError.cjs');
var Message = require('../models/Message.cjs');
var utils = require('../utils.cjs');

function undoFactory(api) {
    const URLType = {
        [Message.MessageType.DirectMessage]: `${api.zpwServiceMap.chat[0]}/api/message/undo?zpw_ver=${context.appContext.API_VERSION}&zpw_type=${context.appContext.API_TYPE}`,
        [Message.MessageType.GroupMessage]: `${api.zpwServiceMap.group[0]}/api/group/undomsg?zpw_ver=${context.appContext.API_VERSION}&zpw_type=${context.appContext.API_TYPE}`,
    };
    /**
     * Undo a message
     *
     * @param message Message or GroupMessage instance that has quote to undo
     *
     * @throws ZaloApiError
     */
    return async function undo(message) {
        var _a;
        if (!context.appContext.secretKey)
            throw new ZaloApiError.ZaloApiError("Secret key is not available");
        if (!context.appContext.imei)
            throw new ZaloApiError.ZaloApiError("IMEI is not available");
        if (!context.appContext.cookie)
            throw new ZaloApiError.ZaloApiError("Cookie is not available");
        if (!context.appContext.userAgent)
            throw new ZaloApiError.ZaloApiError("User agent is not available");
        if (!(message instanceof Message.Message) && !(message instanceof Message.GroupMessage))
            throw new ZaloApiError.ZaloApiError("Expected Message or GroupMessage instance, got: " + ((_a = message === null || message === void 0 ? void 0 : message.constructor) === null || _a === void 0 ? void 0 : _a.name));
        if (!message.data.quote)
            throw new ZaloApiError.ZaloApiError("Message does not have quote");
        const params = {
            msgId: message.data.quote.globalMsgId,
            clientId: Date.now(),
            cliMsgIdUndo: message.data.quote.cliMsgId,
        };
        if (message instanceof Message.GroupMessage) {
            params["grid"] = message.threadId;
            params["visibility"] = 0;
            params["imei"] = context.appContext.imei;
        }
        else
            params["toid"] = message.threadId;
        const encryptedParams = utils.encodeAES(context.appContext.secretKey, JSON.stringify(params));
        if (!encryptedParams)
            throw new ZaloApiError.ZaloApiError("Failed to encrypt message");
        const response = await utils.request(URLType[message.type], {
            method: "POST",
            body: new URLSearchParams({
                params: encryptedParams,
            }),
        });
        const result = await utils.handleZaloResponse(response);
        if (result.error)
            throw new ZaloApiError.ZaloApiError(result.error.message, result.error.code);
        return result.data;
    };
}

exports.undoFactory = undoFactory;
